src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -378,12 +446,13 @@ BookStep<TIn, TOut>::revImp (
       Return true to continue to receive offers, false to stop receiving offers.
     */
     auto eachOffer =
-        [&](TOffer<TIn, TOut>& offer,
-            TAmounts<TIn, TOut> const& ofrAmt,
-            TAmounts<TIn, TOut> const& stpAmt,
-            TOut const& ownerGives,
-            std::uint32_t transferRateIn,
-            std::uint32_t transferRateOut) mutable -> bool
+        [this, &sb, &out, &result, &remainingOut, &savedIns, &savedOuts]
```

Comment:
I disagree with Scott Meyer on this one. I think of a lambda as part of the enclosing function. Its a convenience that lets you essentially create "sub-functions". The idea is that they are somewhat encapsulated and therefore, easier to reason about. And they make the enclosing function similarly easier to reason about. Here's a perfect example of the use of a lambda for that sort of thing (at the end of this message).

So, what about the capture mode? I find that there are usually two types of lambdas. Pure functions that operate only on their inputs (function parameters), and those that mutate variables in the enclosing function scope. For the first kind, they don't need to capture in the first place so its a moot point. For the second kind, I don't see much gain from explicitly listing each individual parameter. You need what you need, no more and no less. Importing more than you need doesn't cost anything and its a convenient shorthand. Quite frankly the lambda shown above, in BookStep.cpp, has a ridiculous list of captures. I agree with Scott's assessment - one capture is probably okay but in this case, its just noise.

Listing:
```
template<class _>
bool
utf8_checker_t<_>::write(std::uint8_t const* in, std::size_t const size)
{
    auto const valid =
        [](std::uint8_t const*& in)
        {
            if (*in < 128)
            {
                ++in;
                return true;
            }
            if ((in[0] & 0x60) == 0x40)
            {
                if ((in[0] & 0xfe) == 0xc0 ||
                    (in[1] & 0xc0) != 0x80)
                        return false;
                in += 2;
                return true;
            }
            if ((in[0] & 0xf0) == 0xe0)
            {
                if ((in[1] & 0xc0) != 0x80 ||
                    (in[2] & 0xc0) != 0x80 ||
                    (in[0] == 224 && in[1] < 160) ||
                    (in[0] == 237 && in[1] > 159))
                        return false;
                in += 3;
                return true;
            }
            if ((in[0] & 0xf8) == 0xf0)
            {
                if (in[0] > 244 ||
                    (in[1] & 0xc0) != 0x80 ||
                    (in[2] & 0xc0) != 0x80 ||
                    (in[3] & 0xc0) != 0x80 ||
                    (in[0] == 240 && in[1] < 144) ||
                    (in[0] == 244 && in[1] > 143))
                        return false;
                in += 4;
                return true;
            }
            return false;
        };
    auto const needed =
        [](std::uint8_t const in)
        {
            if (in < 128)
                return 1;
            if (in < 194)
                return 0;
            if (in < 224)
                return 2;
            if (in < 240)
                return 3;
            if (in < 245)
                return 4;
            return 0;
        };

    auto const end = in + size;
    if (need_ > 0)
    {
        auto n = std::min(size, need_);
        need_ -= n;
        while(n--)
            *p_++ = *in++;
        if(need_ > 0)
        {
            BOOST_ASSERT(in == end);
            return true;
        }
        std::uint8_t const* p = &have_[0];
        if (! valid(p))
            return false;
        p_ = have_;
    }

    auto const last = in + size - 3;
    while(in < last)
        if(! valid(in))
            return false;

    for(;;)
    {
        auto n = end - in;
        if(! n)
            break;
        auto const need = needed(*in);
        if (need == 0)
            return false;
        if(need <= n)
        {
            if(! valid(in))
                return false;
        }
        else
        {
            need_ = need - n;
            while(n--)
                *p_++ = *in++;
            break;
        }
    }
    return true;
}
```